File: code/ex216.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> cur = new ArrayList<>();
    
    public List<List<Integer>> combinationSum3(int k, int n) {
        backtracking(k, n, 1, 0);
        return res;
    }
    
    private void backtracking(int k, int target, int s, int sum) {
        if (sum == target && cur.size() == k) {
            res.add(new ArrayList<>(cur));
            return;
        }
        if (sum > target || cur.size() >= k)    return;
        
        for (int i = s; i <= 9 && sum + i <= target ; i++) {
            cur.add(i);
            backtracking(k, target, i + 1, sum + i);
            cur.remove(cur.size() - 1);
        }
    }
}

File: code/ex216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> cur;
        dfs(res, cur, k, n, 1);
        return res;
    }
    
private:
    void dfs(vector<vector<int>>& res, vector<int>& cur, int k, int n, int s) {
        if (k == 0) {
            if (n == 0)    res.push_back(cur);
            return;
        }
        
        for (int i = s; i <= 9; i++) {
            cur.push_back(i);
            dfs(res, cur, k - 1, n - i, i + 1);
            cur.pop_back();
        }
    }
};